Support Graceful Stopping of Child Routees

michaeljmarshall opened the pull request:
## Purpose

Allow end users to choose which message a router sends to a child routee when the router decreases the number of routees. This change enables graceful stopping of child routees, thus providing support for stateful routees.

## References

References: None known

## Changes

- Add abstract member `routeeStopMessage` to `RouterConfig` trait.
- Adjust `RoutedActorCell` to send the `routeeStopMessage` when stopping child routees.
- Define the `routeeStopMessage` to throw exceptions in the `Group` and `NoRouter` traits since these traits have routees which are, by definition, not children of the router.
- Add `routeeStopMessage` argument to all router apply methods which extend the `Pool` trait since routees are, by definition, children of those routers. In order to maintain the original functionality, and prevent breaking changes, the argument is defaulted to `PoisonPill` in all cases.
- Define `routeeStopMessage` to `PoisonPill` in the `CustomRouterConfig` in order to maintain original functionality. Member is overridable.

## Background Context

For routers managing stateful routees, it would be helpful to enable routees to gracefully stop themselves. The current functionality is simply to send a `PoisonPill` to routees that are children and that need to be stopped. The proposed changes allow for end users to override the default message, which is still `PoisonPill`, without altering the current functionality in any way. Specific use cases include pool routers with a resizer, the `ClusterRouterPool`, and any time a pool router receives the `RemoveRoutee` message. In all three cases, the router can stop routees at any time. As such, it becomes tricky to maintain any type of state in routees. The proposed changes enable routees to maintain state more safely by giving the routees a chance to persist their state.

Since the length of actor shutdown is immaterial to the router's ability to function, the router should not need to waste time ensuring the successful shutdown of the child routee. As such, the router sends the `routeeStopMessage` once and expects the routee to stop itself.

The proposed changes were inspired by the `handOffStopMessage` defined for `ClusterSharding`. 

The original functionality is maintained for end users, and this proposal should not introduce any breaking changes.

## Remaining Work

If the proposed changes are accepted, I will need help determining where to add documentation and understanding testing norms for the akka project.

michaeljmarshall committed:
Add routeeStopMessage to RouterConfig and sub classes

lightbend-cla-validator commented:
Hi @michaeljmarshall,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

johanandren commented:
How is this improving on just sending a custom stop message to the all routees using `Broadcast` (a pool will stop itself if all children stop)?

raboof commented:
OK TO TEST

michaeljmarshall commented:
The improvement does not come when shutting down the whole router. As you demonstrated, it is trivial to gracefully shutdown a router. My proposed changes improve routers that remove some subset of the overall routee pool while in flight. This happens whenever a router receives a `RemoveRoutee` message or an `AdjustPoolSize(x)` message where `x` is a negative integer. A router could receive those messages if the end user sends either message to the router or if it has a `Resizer`. Additionally, cluster routers trigger the removal of some routees when receiving cluster notifications for a node leaving the cluster. In all of those cases, the current implementation is that a router sends a `PoisonPill` to the child routee(s) being removed. My proposed changes improve routers by allowing the end user to configure the message sent to routee(s) being removed as a result of any of the above events.
 
In the case of the cluster router, this change will improve how a cluster router reacts to a node leaving the cluster. Based on the current implementation, when a node registers as leaving the cluster, the cluster router sends a `PoisonPill` to each routee on the node leaving the cluster. If a `PoisonPill` is too blunt, the safest way I have been able to find to ensure routees on the leaving node are gracefully stopped is to stop all routees by broadcasting a stop message and then to have the router bring routees back up once the node in question has successfully left the cluster. However, this approach interrupts all of the routees, which is unnecessary and could lead to lost messages since the router does not have any routees for a time. By adding the `routeeStopMessage`, it makes it possible to ensure that actors are appropriately shutdown without unnecessarily interrupting the whole cluster.

In the case of a router with a resizer, there is no way to gracefully stop routees because the resizer triggers changes to the size of the routee pool independently from the end user. As such, the current implementation prevents the graceful stopping of routees while also using a resizer. The proposed changes increase the versatility of routers with resizers.

An additional benefit of the `routeeStopMessage` is to give end users the ability to account for their tolerance to dropped messages. The current implementation is that the router sends a `PoisonPill` after 100 milliseconds. The delay reduces, but does not eliminate, the chance for dropped messages. However, with this change, an end user could delay the shutdown of the routee to further decrease (although still not eliminate) the likelihood of dropped messages. The end user might do this by scheduling another stop message to self for some time in the future. I could see an end user doing this if they have a low tolerance for dropped messages and are able to take on the addition cost of waiting to shutdown the routee.

Note that this change does not affect the way a router shuts down. If a router is killed by any means (eg `PoisonPill`, `Kill`), it will send system messages to its children to stop them. This proposal only changes the behavior of removing individual routees from the collection of routees.

michaeljmarshall commented:
I realize now that my proposed changes are most likely disqualified because they are not binary compatible. They fail binary compatibility because I added an argument to several class constructors. I had thought that because I provided a default value each time I added new arguments to the class constructors, I would not be introducing any breaking changes; however, based on closer reading about Java's binary compatibility, it looks like that was an incorrect assumption. I don't currently see a way to introduce my proposed changes without also changing the class constructors. What does this mean for my pull request? (This is my first time submitting a pull request to an open source project, so I appreciate any direction here.) Thank you.

johanandren commented:
Ah, sorry, I didn't realise that aspect of it. Sounds like it could makes sense. It would likely require quite a bit of wiggling to keep binary compatibility (it's not impossible though). As we are moving our main focus to 2.6 and typed, I wonder if this effort on improving is better spent on typed routers instead (although the scope of those are much much more limited than untyped routers and I'm not sure we want to aim at where we have ended up with untyped routers at all, we are not going to support remote deployment for example so pools will always be local).

michaeljmarshall commented:
I am closing this PR, as I don't believe there is an easy way to accomplish the desired end without introducing a breaking change. After researching Scala's requirements for class constructors, it appears that the compiler does not allow a class to have multiple constructors with default arguments.
